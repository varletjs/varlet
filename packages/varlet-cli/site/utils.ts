import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
