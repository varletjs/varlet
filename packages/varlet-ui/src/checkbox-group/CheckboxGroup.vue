<template>
  <div :class="n('wrap')">
    <div :class="classes(n(), n(`--${direction}`))">
      <slot />
    </div>
    <var-form-details :error-message="errorMessage" />
  </div>
</template>

<script lang="ts">
import VarFormDetails from '../form-details'
import { defineComponent, computed, watch, nextTick } from 'vue'
import { props } from './props'
import { useValidation, createNamespace, call } from '../utils/components'
import { useCheckboxes } from './provide'
import { useForm } from '../form/provide'
import { uniq } from '../utils/shared'
import type { ComputedRef } from 'vue'
import type { ValidateTriggers } from './props'
import type { CheckboxGroupProvider } from './provide'

const { n, classes } = createNamespace('checkbox-group')
export default defineComponent({
  name: 'VarCheckboxGroup',
  components: { VarFormDetails },
  props,
  setup(props) {
    const max: ComputedRef<number | string | undefined> = computed(() => props.max)
    const checkedCount: ComputedRef<number> = computed(() => props.modelValue.length)
    const { length, checkboxes, bindCheckboxes } = useCheckboxes()
    const { bindForm } = useForm()
    const {
      errorMessage,
      validateWithTrigger: vt,
      validate: v,
      // expose
      resetValidation,
    } = useValidation()
    const checkboxGroupErrorMessage: ComputedRef<string> = computed(() => errorMessage.value)

    const validateWithTrigger = (trigger: ValidateTriggers) => {
      nextTick(() => {
        const { validateTrigger, rules, modelValue } = props
        vt(validateTrigger, trigger, rules, modelValue)
      })
    }

    const change = (changedModelValue: any) => {
      call(props['onUpdate:modelValue'], changedModelValue)
      call(props.onChange, changedModelValue)
      validateWithTrigger('onChange')
    }

    const onChecked = (changedValue: any) => {
      const { modelValue } = props

      if (!modelValue.includes(changedValue)) {
        change([...modelValue, changedValue])
      }
    }

    const onUnchecked = (changedValue: any) => {
      const { modelValue } = props

      if (!modelValue.includes(changedValue)) {
        return
      }

      change(modelValue.filter((value) => value !== changedValue))
    }

    const syncCheckboxes = () => checkboxes.forEach(({ sync }) => sync(props.modelValue))

    const resetWithAnimation = () => {
      checkboxes.forEach((checkbox) => checkbox.resetWithAnimation())
    }

    // expose
    const checkAll = () => {
      const checkedValues: any[] = checkboxes.map(({ checkedValue }) => checkedValue.value)
      const changedModelValue: any[] = uniq(checkedValues)

      resetWithAnimation()

      call(props['onUpdate:modelValue'], changedModelValue)

      return changedModelValue
    }

    // expose
    const inverseAll = () => {
      const checkedValues: any[] = checkboxes
        .filter(({ checked }) => !checked.value)
        .map(({ checkedValue }) => checkedValue.value)
      const changedModelValue: any[] = uniq(checkedValues)

      resetWithAnimation()

      call(props['onUpdate:modelValue'], changedModelValue)

      return changedModelValue
    }

    // expose
    const reset = () => {
      call(props['onUpdate:modelValue'], [])
      resetValidation()
    }

    // expose
    const validate = () => v(props.rules, props.modelValue)

    watch(() => props.modelValue, syncCheckboxes, { deep: true })

    watch(() => length.value, syncCheckboxes)

    const checkboxGroupProvider: CheckboxGroupProvider = {
      max,
      checkedCount,
      onChecked,
      onUnchecked,
      validate,
      resetValidation,
      reset,
      errorMessage: checkboxGroupErrorMessage,
    }

    bindCheckboxes(checkboxGroupProvider)
    call(bindForm, checkboxGroupProvider)

    return {
      errorMessage,
      n,
      classes,
      checkAll,
      inverseAll,
      reset,
      validate,
      resetValidation,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../form-details/formDetails';
@import './checkboxGroup';
</style>
