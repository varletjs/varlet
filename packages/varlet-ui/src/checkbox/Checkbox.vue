<template>
  <div class="var-checkbox__wrap" @click="handleClick">
    <div class="var-checkbox">
      <div
        class="var-checkbox__action"
        :class="[
          checked ? 'var-checkbox--checked' : 'var-checkbox--unchecked',
          errorMessage || checkboxGroupErrorMessage ? 'var-checkbox--error' : null,
          formDisabled || disabled ? 'var-checkbox--disabled' : null,
        ]"
        :style="{ color: checked ? checkedColor : uncheckedColor }"
        v-ripple="{ disabled: formReadonly || readonly || formDisabled || disabled || !ripple }"
      >
        <slot name="checked-icon" v-if="checked">
          <var-icon
            class="var-checkbox__icon"
            name="checkbox-marked"
            :class="[withAnimation ? 'var-checkbox--with-animation' : null]"
            :size="iconSize"
            var-checkbox-cover
          />
        </slot>
        <slot name="unchecked-icon" v-else>
          <var-icon
            class="var-checkbox__icon"
            name="checkbox-blank-outline"
            :class="[withAnimation ? 'var-checkbox--with-animation' : null]"
            :size="iconSize"
            var-checkbox-cover
          />
        </slot>
      </div>

      <div
        class="var-checkbox__text"
        :class="[
          errorMessage || checkboxGroupErrorMessage ? 'var-checkbox--error' : null,
          formDisabled || disabled ? 'var-checkbox--disabled' : null,
        ]"
      >
        <slot />
      </div>
    </div>

    <var-form-details :error-message="errorMessage" />
  </div>
</template>

<script lang="ts">
import VarIcon from '../icon'
import VarFormDetails from '../form-details'
import Ripple from '../ripple'
import { defineComponent, ref, computed, watch, nextTick } from 'vue'
import { props } from './props'
import { useValidation } from '../utils/components'
import { useCheckboxGroup } from './provide'
import { useForm } from '../form/provide'
import type { Ref, ComputedRef } from 'vue'
import type { ValidateTriggers } from './props'
import type { CheckboxProvider } from './provide'

export default defineComponent({
  name: 'VarCheckbox',
  directives: { Ripple },
  components: {
    VarIcon,
    VarFormDetails,
  },
  props,
  setup(props) {
    const value: Ref = ref(false)
    const checked: ComputedRef<boolean> = computed(() => value.value === props.checkedValue)
    const checkedValue: ComputedRef<boolean> = computed(() => props.checkedValue)
    const withAnimation: Ref<boolean> = ref(false)
    const { checkboxGroup, bindCheckboxGroup } = useCheckboxGroup()
    const { form, bindForm } = useForm()
    const {
      errorMessage,
      validateWithTrigger: vt,
      validate: v,
      // expose
      resetValidation,
    } = useValidation()

    const validateWithTrigger = (trigger: ValidateTriggers) => {
      nextTick(() => {
        const { validateTrigger, rules, modelValue } = props
        vt(validateTrigger, trigger, rules, modelValue)
      })
    }

    const change = (changedValue: any) => {
      value.value = changedValue

      const { checkedValue, onChange } = props

      props['onUpdate:modelValue']?.(value.value)
      onChange?.(value.value)
      validateWithTrigger('onChange')

      changedValue === checkedValue ? checkboxGroup?.onChecked(checkedValue) : checkboxGroup?.onUnchecked(checkedValue)
    }

    const handleClick = (e: Event) => {
      const { disabled, readonly, checkedValue, uncheckedValue, onClick } = props

      if (form?.disabled.value || disabled) {
        return
      }

      onClick?.(e)

      if (form?.readonly.value || readonly) {
        return
      }

      withAnimation.value = true
      const maximum = checkboxGroup ? checkboxGroup.checkedCount.value >= Number(checkboxGroup.max.value) : false

      if (!checked.value && maximum) {
        return
      }

      change(checked.value ? uncheckedValue : checkedValue)
    }

    const sync = (values: Array<any>) => {
      const { checkedValue, uncheckedValue } = props
      value.value = values.includes(checkedValue) ? checkedValue : uncheckedValue
    }

    const resetWithAnimation = () => {
      withAnimation.value = false
    }

    // expose
    const reset = () => {
      props['onUpdate:modelValue']?.(props.uncheckedValue)
      resetValidation()
    }

    // expose
    const toggle = (changedValue?: any) => {
      const { checkedValue, uncheckedValue } = props

      const shouldReverse = ![checkedValue, uncheckedValue].includes(changedValue)
      if (shouldReverse) {
        changedValue = checked.value ? uncheckedValue : checkedValue
      }

      change(changedValue)
    }

    // expose
    const validate = () => v(props.rules, props.modelValue)

    watch(
      () => props.modelValue,
      (newValue) => {
        value.value = newValue
      },
      { immediate: true }
    )

    const checkboxProvider: CheckboxProvider = {
      checkedValue,
      checked,
      sync,
      validate,
      resetValidation,
      reset,
      resetWithAnimation,
    }

    bindCheckboxGroup?.(checkboxProvider)
    bindForm?.(checkboxProvider)

    return {
      withAnimation,
      checked,
      errorMessage,
      checkboxGroupErrorMessage: checkboxGroup?.errorMessage,
      formDisabled: form?.disabled,
      formReadonly: form?.readonly,
      handleClick,
      toggle,
      reset,
      validate,
      resetValidation,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../ripple/ripple';
@import '../form-details/formDetails';
@import '../icon/icon';
@import './checkbox';
</style>
