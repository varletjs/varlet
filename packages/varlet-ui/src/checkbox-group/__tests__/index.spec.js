import example from '../example'
import CheckboxGroup from '..'
import Checkbox from '../../checkbox'
import VarCheckboxGroup from '../CheckboxGroup'
import VarCheckbox from '../../checkbox/Checkbox'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay } from '../../utils/jest'

test('test checkbox example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test checkbox group plugin', () => {
  const app = createApp({}).use(CheckboxGroup)
  expect(app.component(CheckboxGroup.name)).toBeTruthy()
})

test('test checkbox plugin', () => {
  const app = createApp({}).use(Checkbox)
  expect(app.component(Checkbox.name)).toBeTruthy()
})

test('test checkbox check value', async () => {
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarCheckbox, {
    props: {
      modelValue: false,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  await wrapper.find('.var-checkbox').trigger('click')
  expect(onUpdateModelValue).lastCalledWith(true)
  expect(wrapper.props('modelValue')).toBe(true)

  await wrapper.find('.var-checkbox').trigger('click')
  expect(onUpdateModelValue).lastCalledWith(false)
  expect(wrapper.props('modelValue')).toBe(false)

  wrapper.unmount()
})

test('test checkbox check value with custom value', async () => {
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarCheckbox, {
    props: {
      modelValue: 0,
      uncheckedValue: 0,
      checkedValue: 1,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  await wrapper.find('.var-checkbox').trigger('click')
  expect(onUpdateModelValue).lastCalledWith(1)
  expect(wrapper.props('modelValue')).toBe(1)

  wrapper.unmount()
})

test('test checkbox onClick & onChange', async () => {
  const onClick = jest.fn()
  const onChange = jest.fn()

  const wrapper = mount(VarCheckbox, {
    props: {
      modelValue: false,
      onClick,
      onChange,
    },
  })

  await wrapper.find('.var-checkbox').trigger('click')
  expect(onClick).toHaveBeenCalledTimes(1)
  expect(onChange).lastCalledWith(true)

  wrapper.unmount()
})

test('test checkbox toggle method', async () => {
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarCheckbox, {
    props: {
      modelValue: false,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  wrapper.vm.toggle()
  await delay(16)

  expect(onUpdateModelValue).lastCalledWith(true)
  expect(wrapper.props('modelValue')).toBe(true)

  wrapper.unmount()
})

test('test checkbox disabled', async () => {
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))
  const onClick = jest.fn()
  const onChange = jest.fn()

  const wrapper = mount(VarCheckbox, {
    props: {
      modelValue: false,
      disabled: true,
      onClick,
      onChange,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  await wrapper.find('.var-checkbox').trigger('click')

  expect(onUpdateModelValue).toHaveBeenCalledTimes(0)
  expect(onClick).toHaveBeenCalledTimes(0)
  expect(onChange).toHaveBeenCalledTimes(0)
  expect(wrapper.props('modelValue')).toBe(false)

  wrapper.unmount()
})

test('test checkbox readonly', async () => {
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))
  const onClick = jest.fn()
  const onChange = jest.fn()

  const wrapper = mount(VarCheckbox, {
    props: {
      modelValue: false,
      readonly: true,
      onClick,
      onChange,
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  await wrapper.find('.var-checkbox').trigger('click')

  expect(onUpdateModelValue).toHaveBeenCalledTimes(0)
  expect(onClick).toHaveBeenCalledTimes(1)
  expect(onChange).toHaveBeenCalledTimes(0)
  expect(wrapper.props('modelValue')).toBe(false)

  wrapper.unmount()
})

test('test checkbox with checkbox group', async () => {
  const wrapper = mount({
    components: {
      [VarCheckboxGroup.name]: VarCheckboxGroup,
      [VarCheckbox.name]: VarCheckbox,
    },
    data: () => ({
      value: [],
    }),
    template: `
      <var-checkbox-group v-model="value">
        <var-checkbox :checked-value="1" />
        <var-checkbox :checked-value="2" />
      </var-checkbox-group>
    `,
  })

  await wrapper.find('.var-checkbox').trigger('click')
  expect(wrapper.vm.value).toStrictEqual([1])
  await wrapper.find('.var-checkbox').trigger('click')
  expect(wrapper.vm.value).toStrictEqual([])

  wrapper.unmount()
})

test('test checkbox validation', async () => {
  const onUpdateModelValue = jest.fn((value) => wrapper.setProps({ modelValue: value }))

  const wrapper = mount(VarCheckbox, {
    props: {
      modelValue: false,
      rules: [(v) => v || '您必须勾选'],
      'onUpdate:modelValue': onUpdateModelValue,
    },
  })

  wrapper.vm.validate()
  await delay(16)

  expect(wrapper.find('.var-form-details__message').text()).toBe('您必须勾选')
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.var-checkbox').trigger('click')
  await delay(16)

  expect(wrapper.find('.var-form-details__message').exists()).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.reset()
  await delay(16)
  expect(wrapper.props('modelValue')).toBe(false)

  wrapper.unmount()
})

test('test checkbox group max', async () => {
  const wrapper = mount({
    components: {
      [VarCheckboxGroup.name]: VarCheckboxGroup,
      [VarCheckbox.name]: VarCheckbox,
    },
    data: () => ({
      value: [],
    }),
    template: `
      <var-checkbox-group :max="1" v-model="value">
        <var-checkbox :checked-value="1" />
        <var-checkbox :checked-value="2" />
      </var-checkbox-group>
    `,
  })

  const checkboxes = wrapper.findAll('.var-checkbox')

  await checkboxes[0].trigger('click')
  expect(wrapper.vm.value).toStrictEqual([1])

  await checkboxes[1].trigger('click')
  expect(wrapper.vm.value).toStrictEqual([1])

  wrapper.unmount()
})

test('test checkbox group validation', async () => {
  const wrapper = mount({
    components: {
      [VarCheckboxGroup.name]: VarCheckboxGroup,
      [VarCheckbox.name]: VarCheckbox,
    },
    data: () => ({
      value: [],
    }),
    template: `
      <var-checkbox-group ref="checkboxGroup" :rules="[v => v.length >= 1 || '至少选一个']" v-model="value">
        <var-checkbox :checked-value="1" />
        <var-checkbox :checked-value="2" />
      </var-checkbox-group>
    `,
  })

  const { checkboxGroup } = wrapper.vm.$refs

  checkboxGroup.validate()
  await delay(16)
  expect(wrapper.find('.var-form-details__message').text()).toBe('至少选一个')
  expect(wrapper.html()).toMatchSnapshot()

  checkboxGroup.reset()
  await delay(16)
  expect(wrapper.vm.value).toStrictEqual([])

  await wrapper.find('.var-checkbox').trigger('click')
  await delay(16)

  expect(wrapper.find('.var-form-details__message').exists()).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
})

test('test checkbox group checkAll & inverseAll methods', async () => {
  const wrapper = mount({
    components: {
      [VarCheckboxGroup.name]: VarCheckboxGroup,
      [VarCheckbox.name]: VarCheckbox,
    },
    data: () => ({
      value: [1],
    }),
    template: `
      <var-checkbox-group ref="checkboxGroup" v-model="value">
        <var-checkbox :checked-value="1" />
        <var-checkbox :checked-value="2" />
      </var-checkbox-group>
    `,
  })

  await delay(16)
  const { checkboxGroup } = wrapper.vm.$refs

  checkboxGroup.inverseAll()
  await delay(16)
  expect(wrapper.vm.value).toStrictEqual([2])

  checkboxGroup.checkAll()
  await delay(16)
  expect(wrapper.vm.value).toStrictEqual([1, 2])

  wrapper.unmount()
})

test('test checkbox group layout direction', async () => {
  const wrapper = mount({
    components: {
      [VarCheckboxGroup.name]: VarCheckboxGroup,
      [VarCheckbox.name]: VarCheckbox,
    },
    data: () => ({
      value: [],
    }),
    template: `
      <var-checkbox-group direction="vertical" v-model="value">
        <var-checkbox :checked-value="1" />
        <var-checkbox :checked-value="2" />
      </var-checkbox-group>
    `,
  })

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})
