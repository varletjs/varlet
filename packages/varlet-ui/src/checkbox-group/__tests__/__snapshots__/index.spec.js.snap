// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test checkbox example 1`] = `
"<div class=\\"app-type\\">基本使用</div>
<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
    <div class=\\"var-checkbox__text\\">当前的值: false</div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">设置状态值</div>
<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
    <div class=\\"var-checkbox__text\\">当前的值: 0</div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">修改图标和颜色</div>
<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--unchecked\\" style=\\"color: rgb(233, 158, 180);\\"><i class=\\"var-icon var-icon--set var-icon-heart-half-full\\" style=\\"transition-duration: 0ms; font-size: 24px;\\"></i></div>
    <div class=\\"var-checkbox__text\\">当前的值: false</div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">禁用</div>
<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--unchecked var-checkbox--disabled\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
    <div class=\\"var-checkbox__text var-checkbox--disabled\\">当前的值: false</div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">只读</div>
<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
    <div class=\\"var-checkbox__text\\">当前的值: false</div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">复选框组</div>
<div class=\\"var-checkbox-group__wrap\\">
  <div class=\\"var-checkbox-group var-checkbox-group--horizontal\\">
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\">吃饭</div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\">睡觉</div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2 button\\" type=\\"button\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">全选</div>
      <div class=\\"var-button__overlay\\"></div>
    </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2 button\\" type=\\"button\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">反选</div>
      <div class=\\"var-button__overlay\\"></div>
    </button>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"relation\\">当前的值: []</div>
<div class=\\"app-type\\">复选框字段校验</div>
<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
    <div class=\\"var-checkbox__text\\">当前的值: false</div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">复选框组字段校验</div>
<div class=\\"var-checkbox-group__wrap\\">
  <div class=\\"var-checkbox-group var-checkbox-group--horizontal\\">
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\">吃饭</div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\">睡觉</div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"relation\\">当前的值: []</div>
<div class=\\"space\\"></div>"
`;

exports[`test checkbox group layout direction 1`] = `
"<div class=\\"var-checkbox-group__wrap\\">
  <div class=\\"var-checkbox-group var-checkbox-group--vertical\\">
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\"></div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\"></div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test checkbox group validation 1`] = `
"<div class=\\"var-checkbox-group__wrap\\">
  <div class=\\"var-checkbox-group var-checkbox-group--horizontal\\">
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked var-checkbox--error\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text var-checkbox--error\\"></div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked var-checkbox--error\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text var-checkbox--error\\"></div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <transition-stub>
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__error-message\\">
        <transition-stub>
          <div>至少选一个</div>
        </transition-stub>
      </div>
      <div class=\\"var-form-details__extra-message\\">
        <transition-stub>
          <!--v-if-->
        </transition-stub>
      </div>
    </div>
  </transition-stub>
</div>"
`;

exports[`test checkbox group validation 2`] = `
"<div class=\\"var-checkbox-group__wrap\\">
  <div class=\\"var-checkbox-group var-checkbox-group--horizontal\\">
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--checked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-marked var-checkbox__icon var-checkbox--with-animation\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\"></div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
    <div class=\\"var-checkbox__wrap\\">
      <div class=\\"var-checkbox\\">
        <div class=\\"var-checkbox__action var-checkbox--unchecked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
        <div class=\\"var-checkbox__text\\"></div>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test checkbox validation 1`] = `
"<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--unchecked var-checkbox--error\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-blank-outline var-checkbox__icon\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
    <div class=\\"var-checkbox__text var-checkbox--error\\"></div>
  </div>
  <transition-stub>
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__error-message\\">
        <transition-stub>
          <div>您必须勾选</div>
        </transition-stub>
      </div>
      <div class=\\"var-form-details__extra-message\\">
        <transition-stub>
          <!--v-if-->
        </transition-stub>
      </div>
    </div>
  </transition-stub>
</div>"
`;

exports[`test checkbox validation 2`] = `
"<div class=\\"var-checkbox__wrap\\">
  <div class=\\"var-checkbox\\">
    <div class=\\"var-checkbox__action var-checkbox--checked\\"><i class=\\"var-icon var-icon--set var-icon-checkbox-marked var-checkbox__icon var-checkbox--with-animation\\" style=\\"transition-duration: 0ms;\\" var-checkbox-cover=\\"\\"></i></div>
    <div class=\\"var-checkbox__text\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;
